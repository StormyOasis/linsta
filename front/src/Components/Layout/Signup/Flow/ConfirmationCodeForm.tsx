import React, { useEffect } from "react";
import { styled } from "styled-components";

import { getAccountsSendVerifyNotification, postAccountsAttempt } from "../../../../api/ServiceController";
import StyledInput from "../../../../Components/Common/StyledInput";
import StyledButton from "../../../../Components/Common/StyledButton";
import { useDispatch } from "react-redux";
import { AppDispatch, actions } from "../../../../Components/Redux/redux";

const ConfirmationWrapper = styled.div`
    align-content: stretch;
    align-items: stretch;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    overflow: visible;
    position: relative;
    max-width: 350px;
    padding: 8px 28px;
`;

const EmailImage = styled.span`
    background-image: url('/public/images/email.svg');
    background-repeat: none;
    display: block;
    height: 84px;
    width: 117px;
`;

const ConfirmationInnerWrapper = styled.div`
    align-content: stretch;
    align-items: stretch;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    flex-grow: 0;
    flex-shrink: 0;
    margin-bottom: 8px;
    margin-top: 16px;
    position: relative;
    overflow: visible; 
`;

const Text = styled.span`
    display: block;    
    margin: 0;
    overflow: visible;
    overflow-wrap: break-word;
    position: relative;
    text-align: center;
    text-wrap: wrap;
    word-break: break-word;
`;

const ResendButton = styled.button`
    background-color: ${props => props.theme['colors'].backgroundColor};
    color: ${props => props.theme['colors'].buttonDefaultColor};
    cursor: pointer;
    align-items: flex-start;
    position: relative;
    text-align: center;
    font-weight: 600;
    border: none;
    overflow-wrap: break-word;
    overflow: visible;
    text-wrap: wrap;
`;

const PrevButton = styled.button`
    background-color: ${props => props.theme['colors'].backgroundColor};
    color: ${props => props.theme['colors'].buttonDefaultColor};
    cursor: pointer;
    align-items: flex-start
    position: relative;
    text-align: center;
    font-weight: 600;
    border: none;
    font-size: 14px;
    display: block;
    line-height: 18px;
    overflow-wrap: break-word;
    overflow: visible;
    text-wrap: wrap;
`;

export type ConfirmationCodeFormProps = {
    userName: string;
    fullName: string;
    emailOrPhone: string;
    password: string;
    confirmationCode: string;
    month: number;
    day: number;
    year: number;
    confirmationCode_valid: boolean,
    handleFormChange: any,
    changePage: any,
}

const ConfirmationCodeForm: React.FC<ConfirmationCodeFormProps> = (props: ConfirmationCodeFormProps) => {
    const dispatch = useDispatch<AppDispatch>();
    
    useEffect(() => {
           resendCode(); //initial send
    }, [])

    function resendCode() {
        getAccountsSendVerifyNotification(props.emailOrPhone)
            .then((res: any) => res)
            .catch((err: any) => {
                console.error(err);
            });
    }

    function handleFormChange (event: React.ChangeEvent<HTMLInputElement>) {
        props.handleFormChange(
            event.target.name,
            event.target.value,
            event.target.value.length > 0
        );
    };    

    async function submitForm() {
        const result = await postAccountsAttempt({
            dryRun: false,
            userName: props.userName,
            fullName: props.fullName,
            emailOrPhone: props.emailOrPhone,
            password: props.password,
            confirmCode: props.confirmationCode,
            month: props.month,
            day: props.day,
            year: props.year,
        });

        if(result.status === 200) {                             
            const userName = props.userName;
            const password = props.password;

            dispatch(actions.authActions.login({userName, password}))
        }
    }

    return (
        <>
            <ConfirmationWrapper style={{ alignItems: "center" }}>
                <EmailImage aria-label="Email Confirmation" />
                <ConfirmationInnerWrapper>
                    <Text style={{ fontWeight: 600, margin: 0 }}>
                        Enter Confirmation Code
                    </Text>
                </ConfirmationInnerWrapper>
                <ConfirmationInnerWrapper>
                    <Text style={{ maxWidth: "100%" }}>
                        Enter the confirmation code we sent to {props.emailOrPhone}.
                        <ResendButton onClick={resendCode}>Resend Code</ResendButton>
                    </Text>
                </ConfirmationInnerWrapper>
            </ConfirmationWrapper>
            <ConfirmationWrapper>
                <ConfirmationInnerWrapper style={{ alignItems: "center", margin: "8px 0", padding: "0 40px" }}>
                    <ConfirmationWrapper style={{ margin: 0, padding: 0 }}>
                        <StyledInput style={{
                            borderRadius: "6px", cursor: "text", lineHeight: "30px",
                            padding: "0px 12px", textAlign: "left", textTransform: "none", width: "240px"}}
                            maxLength={8}
                            type="text"
                            name="confirmationCode"
                            value={props.confirmationCode}
                            placeholder="Confirmation Code"
                            onChange={handleFormChange} 
                            isValid={props.confirmationCode_valid} 
                        />
                    </ConfirmationWrapper>
                    <ConfirmationWrapper style={{ width: "100%", margin: 0, padding: "16px 8px 16px 8px" }}>
                        <StyledButton
                            datatestid="submit-signupconfirm"
                            type="button" text="Next" disabled={!props.confirmationCode_valid}
                            style={{ margin: 0 }} onClick={() => { submitForm(); }}>
                        </StyledButton>
                    </ConfirmationWrapper>
                    <ConfirmationInnerWrapper style={{ margin: 0 }}>
                        <PrevButton onClick={() => props.changePage(-1)}>Go Back</PrevButton>
                    </ConfirmationInnerWrapper>
                </ConfirmationInnerWrapper>
            </ConfirmationWrapper>
        </>
    );
}

export default ConfirmationCodeForm;