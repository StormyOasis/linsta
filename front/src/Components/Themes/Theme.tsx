import React from "react";
import { ThemeProvider } from "styled-components";

const defaultTheme = {
    colors: {
        defaultLinkColor: "black",
        defaultTextColor: "black",
        backgroundColor: "white",
        backgroundDisabled: "rgba(0, 0, 0, .6)",
        borderDefaultColor: "rgb(220, 220, 220)",
        borderDarkColor: "rgb(180,180, 180)",
        inputBackgroundColor: "rgb(250, 250, 250)",
        navLinkHoverColor: "rgb(236, 236, 236)",
        buttonTextColorDefault: "white",
        buttonSecondaryTextColorDefault: "black",
        buttonDefaultColor: "rgb(0, 150, 245)",
        buttonDefaultSecondaryColor: "rgb(240, 240, 240)",
        buttonDefaultColorTrans: "rgba(0, 150, 245, .5)",
        buttonSecondaryDefaultColorTrans: "rgba(240, 240, 240, .5)",
        buttonOnHoverColor: "rgb(25, 120, 240)",        
        buttonSecondaryOnHoverColor: "rgb(220, 220, 220)",        
        inputTextColor: "rgb(115,115,115)",
        navLinkTextColor: "rgb(0, 50, 100)",
        cropperAspectBkgnd: "rgba(25,25,25,.6)",
        cropperAspectBkgndNoTrans: "rgb(25,25,25)",
        mediumTextColor: "rgb(120, 120, 120)",  
        mediaSliderButtonColor: "rgba(0,0,0, .6)",
        mediaSliderButtonBkgndColor: "rgba(225,225,225, .4)",
        warningTextColor: "rgb(235, 70, 90)"
    },    

    input: {
        backgroundColor: "#EFEFEF",        
    },

    sizes: {
        sideBarNavWidthDefault: "244px",
        sideBarNavWidthNarrow: "72px",
        sideBarNavBottomHeight: "50px",
        cropperHeight: "275px",
        defaultModalWidth: "600px",
        maxModalWidth: "75vw",
        maxModalHeight: "75vh",
        minModalHeight: "300px",
        minPostTextEditorHeight: "140px",
        maxPostTextEditorHeight: "140px",
        feedPostMinWidth: "400px",
        maxCommentModalContentHeight: "75vh",
        minCommentModalContentHeight: "300px",
    },

    breakpoints: {
        sm: "576",
        md: "768",
        lg: "1280"
    }
};

export const LexicalEditorTheme = {
    ltr: 'ltr',
    rtl: 'rtl',
};

const Theme = ({children}: any) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
};

export default Theme;
