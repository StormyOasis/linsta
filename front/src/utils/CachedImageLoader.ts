import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}