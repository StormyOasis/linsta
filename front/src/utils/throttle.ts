import { useRef } from 'react';

const useThrottle = (callback: Function, delay: number) => {
    const lastCall = useRef(0);

    return (...args: any[]) => {
        const now = new Date().getTime();
        if (now - lastCall.current >= delay) {
            lastCall.current = now;
            callback(...args);
        }
    };
};

export default useThrottle;