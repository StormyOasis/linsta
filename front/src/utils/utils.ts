import sanitizeHtml from 'sanitize-html';

import { HistoryType, Post, Profile } from "../api/types";
import { postSetFollowStatus, postUpdatePost } from '../api/ServiceController';
import { DEFAULT_PFP } from '../api/config';

/**
 * Utility object for managing navigation and history state.
 */
export const historyUtils: HistoryType = {
    navigate: null,
    location: null,
    isServer: true
}

// Regular expressions for validation
const validatePasswordRegex: RegExp = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@.#$!%*?&^_])[A-Za-z\d@.#$!%*?&_]{8,32}$/;
const phoneRegex: RegExp = /(?:([+]\d{1,4})[-.\s]?)?(?:[(](\d{1,3})[)][-.\s]?)?(\d{1,4})[-.\s]?(\d{1,4})[-.\s]?(\d{1,9})/g;
const emailRegex: RegExp = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

/**
 * Checks if a string is a hashtag (starts with '#').
 */
export const isHashtag = (text: string) => text.startsWith('#');

/**
 * Checks if a string is a mention (starts with '@').
 */
export const isMention = (text: string) => text.startsWith('@');

export const validatePassword = (value: string): boolean => {
    return validatePasswordRegex.test(value);
};

export const validateEmailPhone = (value: string): boolean => {
    if (value == null) {
        return false;
    }

    return emailRegex.test(value) || phoneRegex.test(value);
};

/**
 * Checks if a full name contains at least two words.
 */
export const validateFullName = (value: string): boolean => {
    return value?.trim().split(" ").length > 1;
};

/**
 * Validates a URL string. Empty or undefined is considered valid.
 */
export const validateUrl = (text: string | undefined): boolean => {
    if (text == null || text.length === 0) {
        return true; //No url entered should be considered valid
    }

    try {
        new URL(text);
        return true;
    } catch (e) {
        return false;
    }
}

// Supported video file extensions
const validVideoExtensions: string[] = ["avif", "ogm", "wmv", "mpg", "webm", "ogv", "mov", "asx", "mpeg", "mp4", "m4v", "avi"];
/**
 * Checks if a file path has a valid video extension.
 * @throws Error if the path is invalid.
 */
export const isVideoFileFromPath = (path: string): boolean => {
    if (!path || path.trim().length < 4) {
        throw new Error("Invalid path");
    }

    const ext = path.toLowerCase().substring(path.lastIndexOf(".") + 1);

    return validVideoExtensions.includes(ext);
}

/**
 * Checks if a MIME type string indicates a video.
 * @throws Error if the type is invalid.
 */
export const isVideoFileFromType = (type: string): boolean => {
    if (type == null || type.trim().length < 4) {
        throw new Error("Invalid type");
    }

    return type.toLowerCase().includes("video");
}

/**
 * Converts a base64 string to a File blob object.
 */
export const base64ToBlob = (base64String: string, outFileName: string): File | null => {
    if (base64String === null || base64String.length === 0) {
        return null;
    }

    const contentType = base64String.substring(5, base64String.indexOf(';'));
    const data = base64String.substring(base64String.indexOf(",") + 1);

    return base64ToBlobEx(data, contentType, outFileName);
}

/**
 * Converts a base64 string (without prefix) to a File object.
 */
export const base64ToBlobEx = (base64String: string, contentType: string, outFileName: string): File => {
    const sliceSize = 512;
    const byteCharacters = atob(base64String);
    const byteArrays = [];

    for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
        const slice = byteCharacters.slice(offset, offset + sliceSize);
        const byteNumbers = new Array(slice.length);

        for (let i = 0; i < slice.length; i++) {
            byteNumbers[i] = slice.charCodeAt(i);
        }

        byteArrays.push(new Uint8Array(byteNumbers));
    }

    return new File(byteArrays, outFileName, { type: contentType });
}

/**
 * Converts a Blob to a base64 string.
 */
export const blobToBase64 = async (blob: any): Promise<string | ArrayBuffer | null> => {
    blob = await createBlob(blob);

    return new Promise((resolve, _reject) => {
        const reader = new FileReader();
        reader.onloadend = () => resolve(reader.result);
        reader.readAsDataURL(blob);
    });
}

/**
 * Fetches a resource as a Blob.
 */
export const createBlob = async (url: string): Promise<Blob | null> => {
    try {
        const response = await fetch(url);
        const data = await response.blob();
        return data;
    } catch (err) {
        console.log(err);
    }
    return null;
}

/**
 * Extracts the MIME type from a base64 data URL.
 */
export const extractMimeTypeFromBase64 = (data: string): string | null => {
    if (data == null || data.length === 0) {
        return null;
    }

    return data.substring(5, data.indexOf(';') + 1);
}

/**
 * Extracts a frame from a video element as a JPEG blob URL.
 */
export const extractFrameFromVideo = async (video: HTMLVideoElement): Promise<string | null> => {
    if (video == null) {
        return null;
    }

    const canvas = document.createElement("canvas");
    const ctx = canvas.getContext('2d');

    if (ctx == null) {
        return null;
    }

    canvas.height = video.videoHeight || parseInt(video.style.height);
    canvas.width = video.videoWidth || parseInt(video.style.width);

    ctx.drawImage(video, 0, 0);

    return new Promise((resolve, _reject) => {
        canvas.toBlob((file: any) =>
            resolve(URL.createObjectURL(file))
            , 'image/jpeg')
    })
}

/**
 * Returns a human-readable difference between now and the given date.
 */
export const dateDiff = (dateTime: Date): string => {
    if (dateTime == null) {
        return "now";
    }

    const now = new Date().getTime();
    const dateTimeMilli = new Date(dateTime).getTime();
    const diff = new Date(now - dateTimeMilli);
    let diffMilli = diff.getTime();

    const days = Math.floor(diffMilli / 1000 / 60 / 60 / 24);
    diffMilli = diffMilli - (days * 1000 * 60 * 60 * 24);

    const weeks = Math.floor(days / 7.0);
    const hours = Math.floor(diffMilli / 1000 / 60 / 60);
    diffMilli = diffMilli - (hours * 1000 * 60 * 60);
    const mins = Math.floor(diffMilli / 1000 / 60);

    if (weeks >= 1) {
        return weeks + "w";
    }

    if (days >= 1) {
        return days + "d";
    }

    if (hours >= 1) {
        return hours + "h";
    }

    if (mins >= 1) {
        return mins + "m";
    }

    return "now";
}

/**
 * Formats a Date object as a readable string.
 */
export const getDateAsText = (date: Date): string => {
    return new Date(date).toLocaleDateString('en-us', { year: "numeric", month: "long", day: "numeric" });
}

/**
 * Finds a post by ID from a list of posts.
 */
export const getPostFromListById = (postId: string, posts: Post[]): Post => {
    return posts.find(post => post.postId === postId) ?? ({} as any);
}

/**
 * Checks if a post is liked by a given user.
 */
export const isPostLiked = (userName: string, post: Post): boolean => {
    return !!(post?.global?.likes?.some((like: any) => like.userName === userName));
}

/**
 * Toggles the liked state of a post for a user (Only on the front end, no backend call here).
 */
export const togglePostLikedState = (userName: string, userId: string, post: Post): Post => {
    if (post == null || userName == null || userId == null) {
        return post;
    }

    if (post.global.likes == null) {
        post.global.likes = [];
    }

    const index = post.global.likes.findIndex((value: any) => value.userName === userName);
    if (index === -1) {
        // Username is not in the post's like list, so add it
        post.global.likes.push({ userName, userId });
    } else {
        // Remove the username from the post's like list
        post.global.likes.splice(index, 1);
    }

    return post;
}

/**
 * Sanitizes text for safe HTML display and also returns a plain text version.
 */
export const getSanitizedText = (text: string): [string, string] => {
    if (text == null || text.length === 0)
        return [text, text];

    const textHtml = sanitizeHtml(text, {
        allowedTags: ['b', 'i', 'em', 'strong', 'a', 'br', 'sub', 'sup'],
        allowedAttributes: {
            'a': ['href']
        },
        allowedIframeHostnames: []
    }).trim();

    const textStripped = sanitizeHtml(text, {
        allowedTags: [],
        allowedAttributes: {},
        allowedIframeHostnames: []
    }).trim();

    return [textHtml, textStripped];
}

/**
 * Checks if an element with the given ID is vertically overflowed.
 */
export const isOverflowed = (id: string): boolean => {
    const element = document.getElementById(id);

    if (element == null) {
        return false;
    }

    return element.offsetHeight < element.scrollHeight;
}

/**
 * Follows or unfollows a user by calling the backend service.
 */
export const followUser = async (userId: string, followUserId: string, shouldFollow: boolean): Promise<boolean> => {
    if (userId === followUserId) {
        return true;
    }

    const data = {
        follow: shouldFollow,
        userId: userId,
        followId: followUserId
    };

    const result = await postSetFollowStatus(data);

    return result.status === 200;
}

/**
 * Returns the profile picture URL from a Profile object, or a default if missing.
 */
export const getPfpFromProfile = (profile: Profile | null): string => {
    return profile?.pfp?.length ? profile.pfp : DEFAULT_PFP;
}

/**
 * Returns the profile picture URL from a Post object, or a default if missing.
 */
export const getPfpFromPost = (post: Post): string => {
    if (post.user.pfp == null || post.user.pfp.length === 0) {
        return DEFAULT_PFP;
    }

    return post.user.pfp;
}

/**
 * Splits a full name into first, middle, and last names.
 */
export const splitFullName = (fullName: string): { firstName: string, middleNames: string, lastName: string } => {
    const names = fullName.trim().split(' ');
    const firstName = names[0];
    const lastName = names[names.length - 1];
    const middleNames = names.slice(1, names.length - 1).join(' ');

    return { firstName, middleNames, lastName };
}

/**
 * Updates fields on a post and calls onClose with the result.
 */
export const updatePostFields = async (post: Post, fieldsToUpdate: { key: string, value: any }[], onClose: (data: any) => void) => {
    if (fieldsToUpdate == null || fieldsToUpdate.length === 0) {
        return;
    }

    try {
        const newPost: Post = await updatePost(post, fieldsToUpdate);
        // Close the modal
        onClose({ isCommited: true, post: newPost, isDeleted: false });
    } catch (err) {
        console.error(err)
    }
}


/**
 * Updates a post with the given fields and returns the updated post.
 */
export const updatePost = async (post: Post, fieldsToUpdate: { key: string, value: any }[]): Promise<Post> => {
    if (fieldsToUpdate == null || fieldsToUpdate.length === 0) {
        return post;
    }

    try {
        // build the query from the given fieldsToUpdate array
        let fields: any = null;
        if (fieldsToUpdate.length > 0) {
            fields = {};
            fieldsToUpdate.forEach((field) => {
                fields[field.key] = field.value;
            });
        }

        // Call the update post service
        const results = await postUpdatePost({ postId: post.postId, fields });
        if (results == null || results.status !== 200) {
            throw new Error("Error updating post");
        }

        // update the local returned post adding the fields not returned by ES
        results.data.postId = post.postId
        results.data.user.pfp = post.user.pfp;
        results.data.commentCount = post.global.commentCount;
        if (post?.global?.likes != null) {
            results.data.global.likes = [...post?.global?.likes];
        }

        return results.data as Post;
    } catch (err) {
        console.error(err)
    }
    return post;
}

/**
 * Stores a search query (string or Profile) in localStorage under "recentSearches".
 * Prevents duplicates based on string value or Profile.profileId.
 * @param value - The search query to store.
 * @returns The updated array of stored queries.
 */
export const storeSearchQueries = (value: string | Profile): (string | Profile)[] => {
    const key: string = "recentSearches";
    try {
        const storedData = localStorage.getItem(key);
        const parsedData = storedData ? JSON.parse(storedData) : [];

        // Check if value is a duplicate
        const isDuplicate: boolean = parsedData.some((item: string | Profile) => {
            if (typeof item === 'string') {
                return item === value;
            } else {
                return item.profileId === (value as Profile).profileId;
            }
        })

        if (isDuplicate) {
            return parsedData;
        }

        parsedData.push(value);

        localStorage.setItem(key, JSON.stringify(parsedData));
        return parsedData;
    } catch (err) {
        console.error(err);
    }
    return [];
}

/**
 * Removes a search query (string or Profile) from localStorage "recentSearches".
 * @param value - The search query to remove.
 * @returns The updated array of stored queries.
 */
export const removeStoredSearchQuery = (value: string | Profile): (string | Profile)[] => {
    const key: string = "recentSearches";
    let queries: (string | Profile)[] = [];
    try {
        queries = getStoredSearchQueries();
        const filteredQueries = queries.filter((item: string | Profile) => {
            if (typeof item === 'string') {
                return item !== value;
            } else {
                return item.profileId !== (value as Profile).profileId;
            }
        });
        localStorage.setItem(key, JSON.stringify(filteredQueries));
        return filteredQueries;
    } catch (err) {
        return queries;
    }
}

/**
 * Retrieves all stored search queries (strings or Profiles) from localStorage.
 * @returns An array of stored queries.
 */
export const getStoredSearchQueries = (): (string | Profile)[] => {
    const key: string = "recentSearches";
    const item = localStorage.getItem(key);

    if (!item) {
        return [];
    }

    try {
        return JSON.parse(item);
    } catch (err) {
        return [];
    }
}

/**
 * Builds a space-separated string of user names from selected Profile objects.
 * @param selectedProfiles - A record of Profile objects keyed by string.
 * @returns A string of user names separated by spaces.
 */
export const buildCollabSearchText = (selectedProfiles: Record<string, Profile>): string => {
    if (!selectedProfiles) {
        return "";
    }

    // Clear out the text input if there's no results
    let userNames: string = "";

    Object.values(selectedProfiles).forEach((profile: Profile) => {
        userNames = `${userNames}${profile.userName} `;
    });

    return userNames.trim();
}