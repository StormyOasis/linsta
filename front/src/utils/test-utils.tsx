import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RootState } from "../Components/Redux/redux";
import { authInitialState } from '../Components/Redux/slices/auth.slice';
import { modalInitialState } from '../Components/Redux/slices/modals.slice';
import { profileInitialState } from '../Components/Redux/slices/profile.slice';
import { miscInitialState } from '../Components/Redux/slices/misc.slice';
import { configureStore } from '@reduxjs/toolkit';

import Theme from '../Components/Themes/Theme';

export const mockInitialState = (overrides?: Partial<RootState>): RootState => ({
    auth: { ...authInitialState, ...(overrides?.auth ?? {}) },
    modal: { ...modalInitialState, ...(overrides?.modal ?? {}) },
    profile: { ...profileInitialState, ...(overrides?.profile ?? {}) },
    misc: { ...miscInitialState, ...(overrides?.misc ?? {}) },
});

type ExtendedRenderOptions = {
    store?: ReturnType<typeof configureStore>;
    preloadedState?: any;
    route?: string; // ‚Üê allow overriding the initial route
} & Omit<RenderOptions, "queries">;

const dummyAuthReducer = (state = { user: null, isLoading: false, error: null }, _action: any) => state;
const dummyModalReducer = (state = { isOpen: false }, _action: any) => state;
const dummyProfileReducer = (state = { name: '', email: '' }, _action: any) => state;
const dummyMiscReducer = (state = { someData: null }, _action: any) => state;

const renderWithStore = (
    ui: ReactElement,
    {
        preloadedState,
        store = configureStore({
            reducer: {
                auth: dummyAuthReducer,
                modal: dummyModalReducer,
                profile: dummyProfileReducer,
                misc: dummyMiscReducer,
            },
            preloadedState,
        }),
        route = "/",
        ...renderOptions
    }: ExtendedRenderOptions = {}
) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
        return (
            <Provider store={store}>
                <MemoryRouter initialEntries={[route]}>
                    <Theme>{children}</Theme>
                </MemoryRouter>
            </Provider>
        );
    };

    return {
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
        store,
    };
};

export * from "@testing-library/react";
export { renderWithStore };
export type { ExtendedRenderOptions };