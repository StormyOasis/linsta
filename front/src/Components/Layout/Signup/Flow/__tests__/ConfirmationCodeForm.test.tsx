// Mocks
jest.mock("styled-components", () => {
  const React = require("react");
  function fakeStyled(BaseComponent = "div") {
    // This is the actual component returned by styled()
    const Styled = React.forwardRef((props, ref) =>
      React.createElement(BaseComponent, { ...props, ref }, props.children)
    );
    // The function that acts as the tagged template
    function templateFn(...args) { return Styled; }
    // .withConfig should return a tagged template function
    templateFn.withConfig = () => templateFn;
    // Also add .withConfig to the component for chained calls
    Styled.withConfig = () => templateFn;
    // Make templateFn callable as a component (for React)
    Object.setPrototypeOf(templateFn, Styled);
    return templateFn;
  }
  // Proxy to handle styled.div, styled.span, styled(AnyComponent), etc.
  const styled = new Proxy(fakeStyled, {
    get: (target, prop) => fakeStyled(prop),
    apply: (target, thisArg, argArray) => fakeStyled(argArray[0])
  });
  return { __esModule: true, styled };
});

jest.mock("../../../../../Components/Common/StyledInput", () => ({
  __esModule: true,
  default: (props: any) => <input data-testid="mock-input" {...props} />
}));
jest.mock("../../../../../Components/Common/StyledButton", () => ({
  __esModule: true,
  default: (props: any) => (
    <button data-testid={props.datatestid} disabled={props.disabled} onClick={props.onClick}>{props.text}</button>
  )
}));
jest.mock("../../../../../api/ServiceController", () => ({
  getAccountsSendVerifyNotification: jest.fn(() => Promise.resolve({})),
  postAccountsAttempt: jest.fn(() => Promise.resolve({ status: 200 })),
}));
jest.mock("react-redux", () => {
  const actual = jest.requireActual("react-redux");
  return {
    ...actual,
    useDispatch: () => jest.fn(),
  };
});
jest.mock("../../../../../Components/Redux/redux", () => ({
  AppDispatch: {},
}));
jest.mock("../../../../../Components/Redux/slices/auth.slice", () => ({
  loginUser: jest.fn(),
}));

jest.mock("../../../../../Components/Themes/Theme", () => ({
  __esModule: true,
  default: ({ children }: any) => <div data-testid="mock-theme">{children}</div>
}));


import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import ConfirmationCodeForm, { ConfirmationCodeFormProps } from "../ConfirmationCodeForm";
import { renderWithStore } from "../../../../../utils/test-utils";
import '@testing-library/jest-dom';

import Theme from "../../../../../Components/Themes/Theme";
console.log("Theme in test file:", Theme);

const defaultProps: ConfirmationCodeFormProps = {
  userName: "testuser",
  fullName: "Test User",
  emailOrPhone: "test@example.com",
  password: "password123",
  confirmationCode: "123456",
  month: 1,
  day: 1,
  year: 2000,
  confirmationCode_valid: true,
  handleFormChange: jest.fn(),
  changePage: jest.fn(),
};

describe("ConfirmationCodeForm", () => {
  it("renders all fields and matches snapshot", () => {
    const r = renderWithStore(<ConfirmationCodeForm {...defaultProps} />);
    expect(screen.getByText("Enter Confirmation Code")).toBeInTheDocument();
    expect(screen.getByTestId("mock-input")).toBeInTheDocument();
    expect(screen.getByTestId("submit-signupconfirm")).toBeInTheDocument();
    expect(screen.getByText("Go Back")).toBeInTheDocument();
    expect(r.container).toMatchSnapshot();
  });

  it("disables Next button if confirmationCode_valid is false", () => {
    const r = renderWithStore(<ConfirmationCodeForm {...defaultProps} confirmationCode_valid={false} />);
    expect(screen.getByTestId("submit-signupconfirm")).toBeDisabled();
    expect(r.container).toMatchSnapshot();
  });

  it("calls handleFormChange when input changes", () => {
    const handleFormChange = jest.fn();
    renderWithStore(<ConfirmationCodeForm {...defaultProps} handleFormChange={handleFormChange} />);
    fireEvent.change(screen.getByTestId("mock-input"), { target: { name: "confirmationCode", value: "654321" } });
    expect(handleFormChange).toHaveBeenCalledWith("confirmationCode", "654321", true);
  });

  it("calls changePage(-1) when Go Back is clicked", () => {
    const changePage = jest.fn();
    renderWithStore(<ConfirmationCodeForm {...defaultProps} changePage={changePage} />);
    fireEvent.click(screen.getByText("Go Back"));
    expect(changePage).toHaveBeenCalledWith(-1);
  });

  it("calls resendCode when Resend Code is clicked", () => {
    const { getAccountsSendVerifyNotification } = require("../../../../../api/ServiceController");
    renderWithStore(<ConfirmationCodeForm {...defaultProps} />);
    fireEvent.click(screen.getByText("Resend Code"));
    expect(getAccountsSendVerifyNotification).toHaveBeenCalledWith(defaultProps.emailOrPhone);
  });

  it("calls submitForm when Next is clicked", async () => {
    const { postAccountsAttempt } = require("../../../../../api/ServiceController");
    renderWithStore(<ConfirmationCodeForm {...defaultProps} />);
    fireEvent.click(screen.getByTestId("submit-signupconfirm"));
    // Wait for async effect
    expect(postAccountsAttempt).toHaveBeenCalled();
  });
});