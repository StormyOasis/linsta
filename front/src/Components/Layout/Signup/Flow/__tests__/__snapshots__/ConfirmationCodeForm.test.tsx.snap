// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationCodeForm disables Next button if confirmationCode_valid is false 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      style="align-items: center;"
    >
      <span
        aria-label="Email Confirmation"
      />
      <div>
        <span
          style="font-weight: 600; margin: 0px;"
        >
          Enter Confirmation Code
        </span>
      </div>
      <div>
        <span
          style="max-width: 100%;"
        >
          Enter the confirmation code we sent to 
          test@example.com
          .
          <button>
            Resend Code
          </button>
        </span>
      </div>
    </div>
    <div>
      <div
        style="align-items: center; margin: 8px 0px; padding: 0px 40px;"
      >
        <div
          style="margin: 0px; padding: 0px;"
        >
          <input
            data-testid="mock-input"
            maxlength="8"
            name="confirmationCode"
            placeholder="Confirmation Code"
            style="border-radius: 6px; cursor: text; line-height: 30px; padding: 0px 12px; text-align: left; text-transform: none; width: 240px;"
            type="text"
            value="123456"
          />
        </div>
        <div
          style="width: 100%; margin: 0px; padding: 16px 8px 16px 8px;"
        >
          <button
            data-testid="submit-signupconfirm"
            disabled=""
          >
            Next
          </button>
        </div>
        <div
          style="margin: 0px;"
        >
          <button>
            Go Back
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfirmationCodeForm renders all fields and matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      style="align-items: center;"
    >
      <span
        aria-label="Email Confirmation"
      />
      <div>
        <span
          style="font-weight: 600; margin: 0px;"
        >
          Enter Confirmation Code
        </span>
      </div>
      <div>
        <span
          style="max-width: 100%;"
        >
          Enter the confirmation code we sent to 
          test@example.com
          .
          <button>
            Resend Code
          </button>
        </span>
      </div>
    </div>
    <div>
      <div
        style="align-items: center; margin: 8px 0px; padding: 0px 40px;"
      >
        <div
          style="margin: 0px; padding: 0px;"
        >
          <input
            data-testid="mock-input"
            maxlength="8"
            name="confirmationCode"
            placeholder="Confirmation Code"
            style="border-radius: 6px; cursor: text; line-height: 30px; padding: 0px 12px; text-align: left; text-transform: none; width: 240px;"
            type="text"
            value="123456"
          />
        </div>
        <div
          style="width: 100%; margin: 0px; padding: 16px 8px 16px 8px;"
        >
          <button
            data-testid="submit-signupconfirm"
          >
            Next
          </button>
        </div>
        <div
          style="margin: 0px;"
        >
          <button>
            Go Back
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
